lib/bb/parse/parse_py/BBHandler.py

```diff
@@ -1,378 +1,260 @@
+
+__func_start_regexp__    = re.compile(r"(((?P<py>python(?=(\s|\()))|(?P<fr>fakeroot(?=\s)))\s*)*(?P<func>[\w\.\-\+\{\}\$:]+)?\s*\(\s*\)\s*{$" )
+__inherit_regexp__       = re.compile(r"inherit\s+(.+)" )
+__export_func_regexp__   = re.compile(r"EXPORT_FUNCTIONS\s+(.+)" )
+__addtask_regexp__       = re.compile(r"addtask\s+(?P<func>\w+)\s*((before\s*(?P<before>((.*(?=after))|(.*))))|(after\s*(?P<after>((.*(?=before))|(.*)))))*")
+__deltask_regexp__       = re.compile(r"deltask\s+(.+)")
+__addhandler_regexp__    = re.compile(r"addhandler\s+(.+)" )
+__def_regexp__           = re.compile(r"def\s+(\w+).*:" )
+__python_func_regexp__   = re.compile(r"(\s+.*)|(^$)|(^#)" )
+__python_tab_regexp__    = re.compile(r" *\t")
+
+__infunc__ = []
 __inpython__ = False
 __body__   = []
-__bbpath_found__ = 0
 __classname__ = ""
-classes = [ None, ]
+
+cached_statements = {}
 
 def supports(fn, d):
-    localfn = localpath(fn, d)
-    return localfn[-3:] == ".bb" or localfn[-8:] == ".bbclass" or localfn[-4:] == ".inc"
-
-def inherit(files, d):
-    __inherit_cache = data.getVar('__inherit_cache', d) or ""
-    fn = ""
-    lineno = 0
-    for f in files:
-        file = data.expand(f, d)
-        if file[0] != "/" and file[-8:] != ".bbclass":
+    """Return True if fn has a supported extension"""
+    return os.path.splitext(fn)[-1] in [".bb", ".bbclass", ".inc"]
+
+def inherit(files, fn, lineno, d):
+    __inherit_cache = d.getVar('__inherit_cache', False) or []
+    files = d.expand(files).split()
+    for file in files:
+        if not os.path.isabs(file) and not file.endswith(".bbclass"):
             file = os.path.join('classes', '%s.bbclass' % file)
 
-        if not file in __inherit_cache.split():
-            debug(2, "BB %s:%d: inheriting %s" % (fn, lineno, file))
-            __inherit_cache += " %s" % file
-            include(fn, file, d)
-    data.setVar('__inherit_cache', __inherit_cache, d)
-
+        if not os.path.isabs(file):
+            bbpath = d.getVar("BBPATH")
+            abs_fn, attempts = bb.utils.which(bbpath, file, history=True)
+            for af in attempts:
+                if af != abs_fn:
+                    bb.parse.mark_dependency(d, af)
+            if abs_fn:
+                file = abs_fn
+
+        if not file in __inherit_cache:
+            logger.debug("Inheriting %s (from %s:%d)" % (file, fn, lineno))
+            __inherit_cache.append( file )
+            d.setVar('__inherit_cache', __inherit_cache)
+            include(fn, file, lineno, d, "inherit")
+            __inherit_cache = d.getVar('__inherit_cache', False) or []
+
+def get_statements(filename, absolute_filename, base_name):
+    global cached_statements
 
-def handle(fn, d, include = 0):
-    global __func_start_regexp__, __inherit_regexp__, __export_func_regexp__, __addtask_regexp__, __addhandler_regexp__, __infunc__, __body__, __bbpath_found__, __residue__
+    try:
+        return cached_statements[absolute_filename]
+    except KeyError:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that you are using the `KeyError` Exception to handle missing items. We recommend using the [`get`](https://docs.python.org/3/library/stdtypes.html#dict.get) method which can return `None` or a predetermined default if the item is missing, without raising `KeyError`.